### DISCUSSION TITLE
Supporting TailwindCSS natively

### DISCUSSION THREAD

**Discussion by khawarizmus** [vuetifyjs/vuetify#12597] [Ideas]
### Problem to solve
This is supposed to be an open discussion and a suggestion more then a feature request but since [Vuetify Issuer creator](https://issues.vuetifyjs.com/) doesn't yet support discussions, I went ahead and choose a feature request type.

Vuetify currently uses CSS utility class and few props to control the styling of the framework and those are great. However it requires some effort from the team to maintain and evolve and it's a moving part in this project that the team is handling.

### Proposed solution
Since Vuetify is being ported to Vue 3 (not completely from scratch) and the team has been already looking into unifying the naming conventions of the utility class [based on  VueConf Toronto
Talk by @johnleider which was great BTW](https://youtu.be/aXY4upCtiPI?t=1485) and is doing some work around that.

I suggest that the team considers using [TailwindCSS](https://tailwindcss.com/) utilities class in staid of "re-inventing the wheel" and offloading the overhead to tailwind.

The purpose of this issue is to discuss the pros and cons and why not using TailwindCSS but to me the biggest factors are:
- Offloading the load to a very strong project that is fast growing in development and popularity and building an ecosystem around it.
- TailwindCSS is very flexible and comes with a configuration file that would enable a whit-label kinda approach to theming configuration (which again from the VueConf Toronto
Talk I know the team is trying to achieve).
- TailwindCSS also have a plugin system which would enables further customization for missing utility class (i am thinking of a ripple effect class for example).

Finally, the two projects are both exemplary open source projects and would really make a perfect match, it would be really exciting to see that happen!

**Comment by Rolanddoda** [comment]
🆙!

Does anyone knows if v3 is going to add support for tailwind ?

**Comment by khawarizmus** [comment]
Would love to hear what @johnleider @heatherleider @KaelWD and @MajesticPotatoe think about this. it would be helpful to know if the core team is interested to look further into this discussion/proposal :)

**Comment by johnleider** [comment]
Not to dive in on any specifics, it's something we have discussed as a team and with the new version it could theoretically work. We'll have more on this once we hit beta.

**Reply by Rolanddoda** [reply]
Cool. Thanks for the update. The last time I used Vuetify was 2 years ago and I loved it but unfortunately I don't use it since then.
I am planning though, to use it with Vue 3 and having TailwindCSS support is something I really want. So I am gonna sponsor you with the hope that this would be something that you and the Vuetify team will consider.

**Reply by SkyaTura** [reply]
@johnleider was this publicly discussed anywhere? I'm curious about the fact why a relevant subject like this (moving utilities towards tailwind) instead of maintaining such similar stuff is not being talked anywhere.

Not judging the decisions of the technical team at all, I love Vuetify and the way this whole project is structured. Watching the evolution of the issues on this repo over the years in fact made me a better developer, but I'd like to understand how and why these decisions are being made, if possible.

**Reply by SkyaTura** [reply]
Just to be specific, I understand that TW and Vuetify can easily co-exist by settings all up with prefixes and stuff, what I would like to know is why not use TW as baseblocks for the library and take the already quoted advantages of it extensibility.

**Reply by johnleider** [reply]
There's zero chance of us dropping our CSS utilities.

**Reply by SkyaTura** [reply]
I mean, I don't see value on dropping it now as well. I just wanted to understand if there was any technical reasoning back then when this could be a foreseeable future, or just a personal choice of the team. Which is perfectly fine as well.

I'm not sure if I'm touching on sensitive subject, I feel that what I implied may be somehow offensive. I'm genuinely sorry if that's the case.

Anyway, we can consider this thread answered as won't happen. If someone wonder about it in the feature.

**Reply by johnleider** [reply]
No offense taken. @KaelWD has pushed me on this over the years.

**Reply by manuelmeister** [reply]
@johnleider As I understood, you can now [disable the utility classes](https://vuetifyjs.com/en/features/sass-variables/#disabling-utility-classes). Is that correct?

**Reply by johnleider** [reply]
You just linked to it. What do you think?

**Reply by manuelmeister** [reply]
I mean, does this mean, that vuetify won't clash with tailwind?

**Reply by johnleider** [reply]
Sorry for my snarky response. Yes, it won't clash with tailwind.

**Reply by Rolanddoda** [reply]
@johnleider thank you so much for this! This makes me and my team consider Vuetify for the next project. Is there any place where we can see how it works with Tailwind (a repo or a demo) or any more information on what happens to Vuetify components if we disable all utilities?

Is this the same as the "unstyled" components (which btw I am not sure if it's something added or not yet)

**Reply by johnleider** [reply]
Nothing will happen. Vuetify components don't use utility classes for default styling.

**Reply by Rolanddoda** [reply]
@johnleider sounds great! How about the "unstyled" feature in Vuetify, does it exist? I know that it was called something different but I can't remember the name.

**Reply by johnleider** [reply]
Could you expand on what "unstyled" is?

**Reply by Rolanddoda** [reply]
The white-label approach. In a talk you said that the framework will have a default styling for their components but we can turn that off so we can have our own styles but still have all the well-tested functionality of a component.

That was a really exciting feature but in the docs I can only see: "By default, Vuetify components have no color and are white-label in nature" which sounds very different from what was introduced in the talk.
I personally thought that "white-label" is what others refer to as "unstyled" components where you can turn off the default styling and apply your own.

**Reply by Rolanddoda** [reply]
https://youtu.be/aXY4upCtiPI?t=1916

**Reply by hermes85pl** [reply]
> Nothing will happen. Vuetify components don't use utility classes for default styling.

@johnleider, unfortunately this is not correct. Surely at least the blueprint for MD3 breaks, see #18846. Unless by "default styling" you don't mean blueprints 🙈. Still, as an end user, I perceive a blueprint as a variant of default styles, and the documentation doesn't warn about the side effect of breaking blueprints in case you completely disable the utility classes feature.

**Reply by johnleider** [reply]
Fair enough. Updated the docs https://github.com/vuetifyjs/vuetify/commit/488f594df946ed114c0b3b4b681950d72c679c71

**Reply by Splinterjke** [reply]
The components' `evelvation` prop depends on utility classes. So disabling utils breaks some styles in case a non-zero elevation is considered as a default styling for some copmponents. Anyway, elevation is still easy doable by using unocss (or any other css helpers). Of course enabling `layers` is recommended in that case.
![image](https://github.com/user-attachments/assets/47df8ff2-d2eb-45de-ba25-c57044ef8b64)
